notebooks/pytorch/training.ipynb

```diff
@@ -0,0 +1,306 @@
+    "# Create the model\n",
+    "net = SonarDNN().double()\n",
+    "optimizer = optim.SGD(net.parameters(),\n",
+    "                      lr=0.01,\n",
+    "                      momentum=0.5,\n",
+    "                      nesterov=False)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# ---------------------------------------\n",
+    "# Define the training loop\n",
+    "# ---------------------------------------\n",
+    "def train(net, train_loader, optimizer, epoch):\n",
+    "    \"\"\"Create the training loop\"\"\"\n",
+    "    net.train()\n",
+    "    criterion = nn.BCELoss()\n",
+    "    running_loss = 0.0\n",
+    "\n",
+    "    for batch_index, data in enumerate(train_loader):\n",
+    "        features = data['features']\n",
+    "        target = data['target']\n",
+    "\n",
+    "        # zero the parameter gradients\n",
+    "        optimizer.zero_grad()\n",
+    "        # forward + backward + optimize\n",
+    "        outputs = net(features)\n",
+    "        loss = criterion(outputs, target)\n",
+    "        loss.backward()\n",
+    "        optimizer.step()\n",
+    "\n",
+    "        # print statistics\n",
+    "        running_loss += loss.item()\n",
+    "        if batch_index % 6 == 5:  # print every 6 mini-batches\n",
+    "            print('[%d, %5d] loss: %.3f' %\n",
+    "                  (epoch, batch_index + 1, running_loss / 6))\n",
+    "            running_loss = 0.0"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# ---------------------------------------\n",
+    "# Define the testing loop\n",
+    "# ---------------------------------------\n",
+    "def test(net, test_loader):\n",
+    "    \"\"\"Test the DNN\"\"\"\n",
+    "    net.eval()\n",
+    "    criterion = nn.BCELoss()  # https://pytorch.org/docs/stable/nn.html#bceloss\n",
+    "    test_loss = 0\n",
+    "    correct = 0\n",
+    "\n",
+    "    with torch.no_grad():\n",
+    "        for i, data in enumerate(test_loader, 0):\n",
+    "            features = data['features']\n",
+    "            target = data['target']\n",
+    "            output = net(features)\n",
+    "            # Binarize the output\n",
+    "            pred = output.apply_(lambda x: 0.0 if x < 0.5 else 1.0)\n",
+    "            test_loss += criterion(output, target)  # sum up batch loss\n",
+    "            correct += pred.eq(target.view_as(pred)).sum().item()\n",
+    "\n",
+    "    test_loss /= len(test_loader.dataset)\n",
+    "    print('\\nTest set:\\n\\tAverage loss: {:.4f}'.format(test_loss))\n",
+    "    print('\\tAccuracy: {}/{} ({:.0f}%)\\n'.format(\n",
+    "            correct,\n",
+    "            (len(test_loader) * test_loader.batch_size),\n",
+    "            100. * correct / (len(test_loader) * test_loader.batch_size)))"
+   ]
+  },
+  {
+   "cell_type": "code",
```

Comment:
Question: Is there an easy way to add local predictions?

Reply:
Like `gcloud` or just like a simple prediction?

It runs 10 batches and outputs something like:
```
[10,     6] loss: 0.678
[10,    12] loss: 0.694
[10,    18] loss: 0.712
[10,    24] loss: 0.692
[10,    30] loss: 0.659
[10,    36] loss: 0.679
[10,    42] loss: 0.685
[10,    48] loss: 0.700

Test set:
	Average loss: 3.2214
	Accuracy: 111/208 (53%)
```

Reply:
simple prediction no gcloud. (Native PyTorch)

Reply:
Added